### ISSUE TITLE
Read error when multi-byte characters in file path name (Origin: bugzilla #145186)

### ISSUE THREAD

**Issue by doxygen** [doxygen/doxygen#1166]
status NEW severity *major* in component *general* for *---*
Reported in version *1.5.3-SVN* on platform *Other*
Assigned to: Dimitri van Heesch

On 2004-06-30 11:02:26 +0000, Leo Liang wrote:

    Doxygen will report "Error: source D:/vxml_ips/ÃÃ¯ÃÃ´Â»Â¥ÃÂªÃÃ¸ is not a
    readable
    file or directory", when the path contains multi-byte characters, e.g.
    chinese.
    
    The platform is Windows 2000, locale is P.R.China, using GB2312 (simplifed 
    chinese) as local charset.
    
    I'v traced into the source code, found that the promblem is in QT codes.
    QFile::encodeName() at last calls QString::local8Bit() to transform unicode 
    string to non-unicode string, on Windows platform, it calls
    qt_winQString2MB(),
    qt_winQString2MB() use WideCharToMultiByte to convert string into
    CP_ACP(ANSI)
    code page, but not the code page specified in locale setting, so the result
    is
    wrong.
    
    Bug 145066 may be the same promblem.

On 2004-06-30 13:14:49 +0000, Leo Liang wrote:

    After checked more codes, I found this program is not i18n friendly:(
    Thought it uses unicode internal, but only do ANSI/latin1 transform, if the 
    local code page is other than ANSI, error occurs.

On 2004-06-30 17:50:24 +0000, Dimitri van Heesch wrote:

    I do not have the i18n knowledge to fix this I'm afraid. So if you
    could provide a patch I'm happy to include it, otherwise you should
    restrict directories to ansi character only.

On 2004-07-01 00:57:40 +0000, Leo Liang wrote:

    I had tried to fix it, but seems it's not so easy. The problems are in QT's 
    code. Maybe we had to wait for QT to resolve it?

On 2004-07-03 13:28:45 +0000, Dimitri van Heesch wrote:

    *** Bug 145066 has been marked as a duplicate of this bug. ***

On 2004-07-05 13:29:06 +0000, xujin2000@sina.com wrote:

    I have tried the following modification:
    
    1.Add #include <qtextcodec.h> in doxywizard.cpp and config.cpp;
    
    2.In function static bool loadConfig( QString loadFile ),
    I change   if (!Config::instance()->parse(loadFile)) to
      QTextCodec *pCodec = QTextCodec::codecForLocale();
      char lac_file[256];
      strcpy(lac_file,pCodec->fromUnicode(loadFile));
      if (!Config::instance()->parse(lac_file))
    
    3.In function static QCString configFileToString(const char *name),
    I change    QFileInfo fi(name) and    f.setName(name)  to
      QTextCodec *pCodec = QTextCodec::codecForLocale();
      QString strName = pCodec->toUnicode(name);
      QFileInfo fi(strName) and f.setName(strName)
    
    And it seems to work properly on my computer.But I am not sure if my 
    modification is correct. I think modify the function to use QString instead
    of
    const char* as parameter is a better way.
    
    

On 2004-07-06 00:51:35 +0000, xujin2000@sina.com wrote:

    By the way, the above modification just intend to correct the doxywizard's 
    unicode path open error. The code involved is so wide and I'm afraid I can't
    do
    it.

On 2007-10-15 19:40:36 +0000, Dimitri van Heesch wrote:

    This bug's version was set to "latest". Since this is a moving target, I
    changed it to 1.5.3-SVN. If you believe this has already been fixed, then
    please change the status accordingly.

**Comment by Warmyone** [comment]
Is there any news/workaround/progress for this?
I'm getting a similar error:
```
> C:/.../doxygen_1.8.18-src/doxygen.exe -s -g C:/.../build/ci-büld/CMakeDoxyfile.tpl
error: Cannot open file C:/.../build/ci-b▒ld/CMakeDoxyfile.tpl for writing
```
This CMake Issue (https://gitlab.kitware.com/cmake/cmake/-/issues/20957) does contain more information.

**Comment by albert-github** [comment]
To use characters outside the range `[A-Za-z0-9_]` is never a good idea but unfortunately in some countries it is common to do it (not to speak of the countries with non-latin character sets).

when I try to create a Doxyfile with:
```
doxygen -g -s ..\win_ü_rel\Doxyfile
```
I get the message:
```
error: Cannot open file ..\win_³_rel\Doxyfile for writing
```
the directory `..\win_ü_rel` exists and is writable.
Similar happens when using an absolute path.
So something to investigate.

When using Cygwin and try to create the Doxyfile in a similar way the Doxyfile is is created.

@Warmyone 
- how did you create your directory with the umlaut (I used the `<ALT>+0252` (on numeric keypad)
- any locale set?

**Comment by Warmyone** [comment]
Using characters outside of that range is necessary as customers in our region do use them.

I'm from germany so there is a key for `ü` on my keyboard. On win10 you can press `<WIN>+.` to open a UI with special chars. Under "symbols" -> "latin signs" you can find `ü` and others. That way you no not need `<ALT>+0252`.

I have not explicitly set any locale, though my region is germany. I also tried setting the cmd to utf8 with `chcp 65001` before calling doxygen so it will get an utf8 string but I get the same error.

**Comment by albert-github** [comment]
With the UI I get the same message, my normal Active code page: 850
When setting the code page `chcp 65001` I get:
```
error: Cannot open file ..\win_
```

(I use a US-English keyboard and English (United States) version of Windows 10, even though I'm from a non English speaking country)

**Comment by Jiangshan00001** [comment]
this still is a problem...

**Comment by albert-github** [comment]
A fix has been made, see commit e002e29, to fix this issue (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.6.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
